src/crypto/transform.py

```diff
@@ -90,40 +79,130 @@ def date_dim_row(self) -> list:
 
         rows = []
         for row in crypto_data:
-            if row["symbol"] in symbols:
+            if row["symbol"] in existing_symbols:
                 continue
             name_dim = NameDim(name_key=row["name"], symbol=row["symbol"], slug=row["slug"])
             rows.append(name_dim._asdict())
         return rows
 
+    def _get_tags(self, crypto_data: dict, existing_tags: set) -> set:
+        tags = set()
+        for row in crypto_data:
+            for tag in row["tags"]:
+                if tag in existing_tags:
+                    continue
+                tags.add(tag)
+        return tags
+    
+    def tag_dim_rows(self, crypto_data: dict) -> list:
+        existing_tags_query = """SELECT tag FROM {tag_dim}""".format(tag_dim=Table.TAG_DIM.value)
+        result = self.bq_client.query(query=existing_tags_query).result()
+        existing_tags = set(row["tag"] for row in result)
+
+        tags = self._get_tags(crypto_data=crypto_data, existing_tags=existing_tags)
+        
+        max_key = 0
+        if tags and not Project.INITIAL_LOAD.value:
+            tag_query = """SELECT MAX(tag_key) AS max_key FROM {tag_dim}""".format(tag_dim=Table.TAG_DIM.value)
+            result = self.bq_client.query(query=tag_query).result()
+            max_key = [row["max_key"] for row in result][0]
+
+        rows = []
+        start = max_key if max_key else 1        
+        for key, tag in enumerate(tags, start=start):
+            tag_dim = TagDim(tag_key=key, tag=tag)
+            rows.append(tag_dim._asdict())
+        return rows
+
+    def name_tag_bridge_table(self, date_key: date, crypto_data: dict) -> list:
+        tag_query = """SELECT tag_key, tag FROM {tag_dim}""".format(tag_dim=Table.TAG_DIM.value)
+        result = self.bq_client.query(query=tag_query).result()
+        tag_dim = {row["tag_key"]: row["tag"] for row in result}
+
+        rows = []     
+        for row in crypto_data:
+            name = row["name"]
+            tags = row["tags"]
+            if tags:
+                for tag in tags:
+                    tag_key = [key for key, value in tag_dim.items() if value == tag][0]
+                    name_tag = NameTag(name_key=name, date_key=date_key, tag_key=tag_key)._asdict()
+                    rows.append(name_tag)
+        return rows
+    
+    def quote_dim_rows(self, date_key: date, crypto_data: dict) -> list:
+        rows = []
+        for row in crypto_data:
+            quote_dim = QuoteDim(name_key=row["name"], date_key=date_key, quote=[row["quote"]])
+            rows.append(quote_dim._asdict())
+        return rows
+    
+    def price_fact_rows(self, date_key: date, crypto_data: dict) -> list:
+        rows = []
+        for row in crypto_data:
+            price = round(row["quote"]["USD"]["price"], 5)
+            price_fact = PriceFact(name_key=row["name"], date_key=date_key, price=price)
+            rows.append(price_fact._asdict())
```

Comment:
yes, I prefer that too. Reason I've been converting the namedtuples to dicts is because it's requirement to load into BQ , at least the method I'm using. I tried loading with namedtuple but the load was rejected.

Reply:
gotcha, makes sense